src/box/memtx_tx.c

```diff
@@ -1328,61 +1346,141 @@ memtx_tx_history_rollback_stmt(struct txn_stmt *stmt)
+		if ((*dels)->txn == stmt->txn) {
+			/* Leave it in list. */
+			dels = &((*dels)->next_in_del_list);
+		} else {
+			if ((*dels)->does_require_old_tuple)
+				/* Conflict, remove from list. */
+				memtx_tx_handle_conflict(stmt->txn,
+							 (*dels)->txn);
+			/* Rebind to story. */
+			(*dels)->del_story = story;
+			struct txn_stmt *save_next = (*dels)->next_in_del_list;
+			/* 1) Link to story's list. */
+			(*dels)->next_in_del_list = story->del_stmt;
+			story->del_stmt = *(dels);
+			/* 2) Unlink to old_story's list. */
+			*dels = save_next;
+		}
+	} while (*dels != NULL);
+}
+
 void
 memtx_tx_history_prepare_stmt(struct txn_stmt *stmt)
 {
 	assert(stmt->txn->psn != 0);
 
-	/* Move story to the past to prepared stories. */
-
+	/**
+	 * History of a key in an index can consist of several stories.
+	 * The list of stories is started with a dirty tuple that is in index.
+	 * The list begins with several (or zero) of stories that are added by
+	 * in-progress transactions, then the list continues with several
+	 * (or zero) of prepared stories, which are followed by several
+	 * (or zero) of committed stories.
+	 * If a statement becomes prepared, its story must be moved to the
+	 * point in list exactly between all still in-progress and all already
+	 * prepared.
+	 */
 	struct memtx_story *story = stmt->add_story;
 	uint32_t index_count = story == NULL ? 0 : story->index_count;
 	/*
+	 * That's a common loop for both index iteration and sequential push
+	 * of the story down in lists of stories.
 	 * Note that if stmt->add_story == NULL, the index_count is set to 0,
 	 * and we will not enter the loop.
 	 */
 	for (uint32_t i = 0; i < index_count; ) {
-		bool old_story_is_prepared = false;
 		struct memtx_story *old_story = story->link[i].older_story;
 		if (old_story == NULL) {
 			i++;
 			continue;
 		}
-		if (old_story->del_psn != 0) {
-			/* if psn is set, the change is prepared. */
-			old_story_is_prepared = true;
-		} else if (old_story->add_psn != 0) {
-			/* if psn is set, the change is prepared. */
-			old_story_is_prepared = true;
-		} else if (old_story->add_stmt == NULL) {
-			/* ancient. */
-			old_story_is_prepared = true;
-		} else if (old_story->add_stmt->txn == stmt->txn) {
-			/* added by us. */
+
+		/*
+		 * We have a story that overwrites an old_story.
+		 * But there could be other transactions that are about to
+		 * delete the same old_story. If such a transaction requires
+		 * exact tuple to be deleted - it must be aborted. If it just
+		 * deletes a key without specifying the tuple - it must be
+		 * relinked to delete current story.
+		 */
+		if (i == 0)
+			memtx_tx_history_prepare_handle_del_stmt(story);
+
+		struct tx_read_tracker *tracker;
+		rlist_foreach_entry(tracker, &old_story->reader_list,
+				    in_reader_list) {
+			if (tracker->reader == stmt->txn)
+				continue;
+			if (tracker->reader->status != TXN_INPROGRESS)
+				continue;
+			memtx_tx_handle_conflict(stmt->txn,
```

Comment:
memtx_tx_handle_conflict() is called inside memtx_tx_history_prepare_handle_del_stmt(). Why should we call it again?